### ISSUE TITLE
Introduce more sophisticated approval rights

### ISSUE THREAD

**Issue by bkchr** [polkadot-fellows/runtimes#93]
Currently the approval rights allow to only specify a minimum rank and the number of approvals required to approve a pull request. While this is working, it is not "very inviting" as lower ranks are not incentivized to even review/approve. However on the other side it is sometimes hard to "cat herd" to get everyone together for reviewing/approval. Thus, we should improve the approval bot to have some sort of "voting power". This voting power should be similar to how on chain voting is done. Every rank gets a certain voting power and we will then require X amount of voting power to approve a pull request. Thus, we could for example have that 3 rank II members together have the same voting power as one rank III member.

CC @Bullrich

**Comment by bkchr** [comment]
The best would be if the bot would support to declare some kind of lambda function. So, that we can implement the logic here in the repo.

**Comment by Bullrich** [comment]
Hi Basti!

We have been discussing on how to appropriately handle this case.

Instead of a lambda function (which can be a security risk), we were thinking about an array that claims the voting power of each dan.

E.g.:
```yml
dan1: 1
dan2: 2
dan3: 5
dan4: 7
dan5: 10
dan6: 15
```

and each rule has a requirement level of approval.

For example, CI files would have a requirement of 14 points, which could be fulfilled by one approval of dan6 or a different combination.

This _voting power_ would be in the `review-bot.yml`, allowing users to adjust such values to your liking.

Let me know what you think!

**Comment by mordamax** [comment]
@bkchr could you provide a feedback or may be ping other people who could also comment on this?
Thanks

**Comment by bkchr** [comment]
Sounds good, but it should also include the possibility to require that at least X from rank Y have approved.

**Comment by Bullrich** [comment]
Hi @bkchr!

I have been thinking on how to approach the given request.

I have the following proposal:

Let's use my idea from before and say we have a field where we define the value of all the ranks, and then we set up the minimum amount of approvals of a given rank, and the minimum "score" it must have for it to pass.

For example:

```yml
score:
  dan1: 1
  dan2: 2
  dan3: 5
  dan4: 9
  dan5: 10
  dan6: 15
rules:
  - name: CI Files
    condition:
      include:
        - ^\.github/.*
    type: fellows
    minRank: 3
    minApprovals: 1
    minScore: 18
```

This would work in the following way:

`CI Files` would require at least *one* approval from someone who is *dan 3 or greater*. And the combined value of all the approvals, defined in `minScore`, must get to 18.

So, in this case, your vote, as a dan 6, would count 15 points. And we would still need 3 more points. Could either be from:
- 3 `dan 1`.
- 1 `dan2` and 1 `dan1`.
- 2 `dan2`.

Of course you can define the value of each dan as you see fit.

Let me know if you think this will fulfill the requirement you have with enough flexibility so I can get started on it.

**Comment by bkchr** [comment]
Sounds good!  :)

**Comment by Bullrich** [comment]
Perfect.

I have created a related issue to keep track of the development: paritytech/review-bot#110

Once it is ready, I'll update the version in this repo 🙂

**Comment by Bullrich** [comment]
PR #168 has been opened which resolves this issue.

**Comment by Bullrich** [comment]
PR #168 has been merged (932fc1555e9892c66f8394ee55c63faf4436d7a4).

Seeing that it is the second time that the ticket did not close with my PR I'm suspecting that an external contributor's commit can not close a issue.

@bkchr, if the requirements for this issue are satisfied, could you please close it?

Thank!

**Comment by bkchr** [comment]
Yeah! Thank you!
